MNT fix test for binary crossentropy in hist-GBDT

lorentzenchr opened the pull request:
#### What does this implement/fix? Explain your changes.
This fixes a test for the binary cross entropy loss function in the module for `HistGradientBoostingClassifier`.

#### Any other comments?
See added comments in commits.

Commit:
TST fix loss function test in _hist_gradient_boosting

rth commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -68,7 +70,13 @@ def test_derivatives(loss, x0, y_true):
 
     def func(x):
-        return loss.pointwise_loss(y_true, x)
+        if type(loss) == _LOSSES['binary_crossentropy']:
```
```suggestion
        if isinstance(loss, _LOSSES['binary_crossentropy']):
```

rth commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -68,7 +70,13 @@ def test_derivatives(loss, x0, y_true):
+            # Subtract a constant term such that the binary cross entropy
+            # has its minimum at zero. This only works if 0 < y_true < 1.
+            return loss.pointwise_loss(y_true, x) \
+                - loss.pointwise_loss(y_true, logit(y_true))
```
I feel this is a bit complex.

Can we rather write, `
```py
if isinstance(loss, _LOSSES['binary_crossentropy']):
    assert_allclose(loss.pointwise_loss(y_true, optimum), 
                    loss.pointwise_loss(y_true, logit(y_true)))
else:
    ...
```
in the test directly?

lorentzenchr replied:
Good idea. Unfortunately, the `newton` root finder is allergic to such chances and complains with a no converged message as it can't ever reach a true zero. I would have to change from `newton` to a minimizer. I thought my change is the smaller one and the advantage of `newton` is that you also check for gradient and hessian in some way being correct. 
What do you think?

rth commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -78,7 +86,7 @@ def fprime2(x):
     optimum = newton(func, x0=x0, fprime=fprime, fprime2=fprime2)
     assert np.allclose(loss.inverse_link_function(optimum), y_true)
-    assert np.allclose(loss.pointwise_loss(y_true, optimum), 0)
+    assert np.allclose(func(optimum), 0)
```
Not asking to change, but all these should have been `assert_allclose` as that produces nicer traceback on failure I think? With possible adjusted tolerances as they are not the same between `np.allclose` and `assert_allclose`.

NicolasHug replied:
This kind of discussion pops up every once in a while, I've asked, there's no consensus.

https://github.com/scikit-learn/scikit-learn/issues/13180

Honestly, I don't really care anymore.

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -48,10 +49,11 @@ def get_hessians(y_true, raw_predictions):
-    ('binary_crossentropy', -12, 1),
-    ('binary_crossentropy', 30, 1),
+    # The argmin of binary_crossentropy for y_true=±1 is ±inf due to logit,
+    # cf. "complete separation". Therefore we use 0 < y_true < 1.
```
you mean 0 / 1 in stead of -1 / 1?

Also, why did it work with y_true == 1 then?

lorentzenchr replied:
Yes, I meant y = 0 gives -inf, y = 1 gives +inf.
The reason, it worked for `y_true = 1` so far was ... luck with numerical precision.

I looked at the solution of `newton`:
- For `y_true = 1` I find `optimum = 35.47761173`. Then `pointwise_loss` calls `np.logaddexp(0, optimum) - optimum` which gives `0` (true for all values of `optimum>33.271...`). 
- Then, for `y_true = 0`, the loss becomes `np.logaddexp(0, x)` which equals zero for `x<-745.1...`, which the Newton/Halley root finder does not find.

Commit:
TST address review comments

Commit:
TST replace allclose with assert_allclose

Commit:
TST adjust tolerances

lorentzenchr commented:
I just rebased. Merge conflicts resolved. CI lights are green.
@rth @NicolasHug I addressed your comments.

PS: As you might note, this is a very subtle, small step towards thinking of logistic regression as a _regression_ on the domain [0, 1]:smirk:

rth commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -81,9 +89,9 @@ def fprime2(x):
-    assert np.allclose(loss.pointwise_loss(y_true, optimum), 0)
-    assert np.allclose(get_gradients(y_true, optimum), 0, atol=1e-7)
+    assert_allclose(loss.inverse_link_function(optimum)[0], y_true)
+    assert_allclose(func(optimum)[0], 0, atol=1e-14)
```
Just a question on superficial read: doesn't `func` return a scalar (I'm a bit confused by the indexing) and if it's doesn't (and returns a ndarray as `loss.pointwise_loss`) why does using newton minimizer with it works as an objective function ?

Maybe some light type annotations on `func` could have helped. mypy now runs as part of CI..

lorentzenchr replied:
🤔 No, `func` returns an array with one element. I tried to return scalars everywhere which fails badly because the methods of loss classes often call `reshape(-1)` internally. (And I have not yet completely figured out why, but this is another story).

I added a comment. I can also add type annotations. Can you provide an short example with ndarrays (of one element)?

rth replied:
> Can you provide an short example with ndarrays (of one element)?

The typing for ndarray dimensions is not too standardized yet (https://github.com/numpy/numpy-stubs), but I think even,
```py
def func(x: np.ndarray) -> np.ndarray:
```
would have been useful as opposed to,
```py
def func(x: np.ndarray) -> float:
```

but anyway it's just a side comment,

rth reviewed (commented):
One question otherwise LGTM, thanks!

Commit:
TST add comment why 1-element arrays instead of scalars

rth reviewed (approved):
Thanks!

Commit:
TST add type annotations and work with scalars

lorentzenchr commented:
@rth Thanks for the approval and the annotation hint. I was able to make it work mostly with scalars. I hope it is now easier to read (and your approval still holds, maybe the type annotations are now too much?)

rth commented:
Thanks! The current version (even without annotations) looks good to me.

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -70,20 +72,26 @@ def test_derivatives(loss, x0, y_true):
-    def fprime2(x):
-        return get_hessians(y_true, x)
+    def fprime2(x: float) -> float:
+        return get_hessians(y_true, x).item(0)
```
instead of dealing with indices here (and below), maybe we should instead do this just before the assertsions:

```py
    y_true = y_true.ravel()
    optimum = optimum.ravel()
```

lorentzenchr replied:
See comment above. AFAIU, the `newton` root finder prefers scalars.
Maybe, I can't see the forest for the trees: `np.array([1.]).ravel()` still returns a ndarray.

NicolasHug replied:
newton works fine with arrays (as long as they have one element like here).

> np.array([1.]).ravel() still returns a ndarray.

Yes, but they would all have the same shape, and it would make `assert_allclose` happy.

The changes you had to do with the use of `item()` and stuff come from the fact that you're now using `assert_allclose` which cannot  compare arrays with different dimensions. It's not related to newton.

lorentzenchr replied:
I see. 👍

NicolasHug replied:
lol yeah I had to do a few back and forth with master myself to really understand what was going on

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -70,20 +72,26 @@ def test_derivatives(loss, x0, y_true):
+        if isinstance(loss, _LOSSES['binary_crossentropy']):
+            # Subtract a constant term such that the binary cross entropy
+            # has its minimum at zero. This only works if 0 < y_true < 1.
+            return (loss.pointwise_loss(y_true, x)
+                    - loss.pointwise_loss(y_true, logit(y_true))).item(0)
```
may I suggest 

```py
            actual_min = loss.pointwise_loss(y_true, logit(y_true))
            return loss.pointwise_loss(y_true, x) - actual_min
```

NicolasHug reviewed (commented):
Thanks @lorentzenchr , a few comments regading readability but LGTM

rth commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -70,20 +72,26 @@ def test_derivatives(loss, x0, y_true):
 
-    def fprime(x):
-        return get_gradients(y_true, x)
+    def fprime(x: float) -> float:
```
> I guess we need a more global decision on whether to use type annotations

I think the current situation is outlined in https://github.com/scikit-learn/scikit-learn/blame/fb76de72e7560aaa739872e94bee761777e54c0a/doc/developers/contributing.rst#L426

I did mention it in a dev meeting a month or two ago and didn't hear much objections. Again the idea is not to type everything, but now and then it helps readability. I think we can keep these in, they can't hurt in any case.

rth replied:
Here in particular I find that without type annotations (and docstrings), it's very difficult to know, what's the expected input and output of,
```py
    def fprime(x):
        return get_gradients(y_true, x).item(0)
```

NicolasHug replied:
Ok.

In this specific case, I think we should actually let numpy deal with arrays in these functions (comment just below), so we'd need to remove the annotations anyway

NicolasHug replied:
Also, the input is actually not a float, it's a 2d array (with only one element).

I really think we should just let numpy and scipy properly deal with broadcasting here, and only ravel the arrays at the very end.

lorentzenchr replied:
The argument `x` of `def func(x)` is the `x0` of `@pytest.mark.parametrize('loss, x0...)` and is actually a float and not a ndarray. Same for `fpime` and `fprime2`.
AFAIU, the `newton` root finder does prefer scalars, not ndarrays as those have different meaning.

I'll do whatever you prefer here. It's just not clear to me what that is 😏

NicolasHug replied:
> AFAIU, the newton root finder does prefer scalars

That's true, but it does fine with arrays with a single element

We need to wrap x0 and y_true into arrays because that's what the loss methods expect. `newton` is OK with these. If we're going to recast these 1-element arrays into floats, it's cleaner and easier to do it once at the very end, i.e. by calling `.ravel` just before the assertions.

Alternatively, we can also go back to just using `assert np.allclose`. I know we sort of prefer `assert_allclose`, but at the end of the day, that change is unrelated to this PR.

Commit:
TST add actual_min for readability

Commit:
TST use ndarray and correct type annotation

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -70,20 +72,28 @@ def test_derivatives(loss, x0, y_true):
-    assert np.allclose(loss.inverse_link_function(optimum), y_true)
-    assert np.allclose(loss.pointwise_loss(y_true, optimum), 0)
-    assert np.allclose(get_gradients(y_true, optimum), 0, atol=1e-7)
+    y_true = y_true.ravel()
```
```suggestion

    # Need to ravel arrays because assert_allclose requires matching dimensions
    y_true = y_true.ravel()
```

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -70,20 +72,28 @@ def test_derivatives(loss, x0, y_true):
+            # Subtract a constant term such that the binary cross entropy
+            # has its minimum at zero. This only works if 0 < y_true < 1.
+            actual_min = loss.pointwise_loss(y_true, logit(y_true))
+            return (loss.pointwise_loss(y_true, x) - actual_min)
```
```suggestion
            return loss.pointwise_loss(y_true, x) - actual_min
```

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -70,20 +72,28 @@ def test_derivatives(loss, x0, y_true):
+            return loss.pointwise_loss(y_true, x)
 
-    def fprime(x):
+    def fprime(x: np.ndarray) -> np.ndarray:
```
I don't feel like we need the annotations anymore. Especially given that these expect arrays with a single element. Newton  would break otherwise (I think). Maybe a comment is enough?

Commit:
TST improve comments in test code

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -61,29 +64,41 @@ def get_hessians(y_true, raw_predictions):
+    def func(x: np.ndarray) -> np.ndarray:
+        if isinstance(loss, _LOSSES['binary_crossentropy']):
+            # Subtract a constant term such that the binary cross entropy
+            # has its minimum at zero. This only works if 0 < y_true < 1.
```
```suggestion
            # has its minimum at zero, which is needed for the newton method.
            # This only works if 0 < y_true < 1.
```

NicolasHug reviewed (approved):
thanks @lorentzenchr , one last nit

I still think we don't need the anotations, but I'll merge tomorrow anyway

Commit:
TST address review comment

NicolasHug commented:
thanks @lorentzenchr !
